src/packages/storage-provider/src/storageProvider.ts

```diff
@@ -0,0 +1,150 @@
+import {
+	GetObjectCommand,
+	PutObjectCommand,
+	S3Client,
+	DeleteObjectCommand,
+} from '@aws-sdk/client-s3';
+import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
+import { graphweaverMetadata, BaseContext } from '@exogee/graphweaver';
+import { GraphQLResolveInfo, Source } from 'graphql';
+import { GraphQLJSON } from '@exogee/graphweaver-scalars';
+import { randomUUID } from 'crypto';
+import { MediaType } from './decorators/media-field';
+
+export enum StorageType {
+	S3 = 's3',
+}
+
+type StorageConfig = {
+	type: StorageType;
+	bucketName: string;
+	region?: string;
+	expiresIn?: number;
+	endpoint?: string;
+};
+
+const EXPIRE_TIME = 3600;
+
+const getMediaType = (ext: string) => {
+	// if image extension return image type
+	if (['png', 'jpeg', 'jpg', 'gif'].includes(ext)) {
+		return MediaType.IMAGE;
+	}
+	return MediaType.OTHER;
+};
+
+export class S3StorageProvider {
+	bucketName: string;
+	region: string | undefined;
+	expiresIn: number;
+	endpoint?: string;
+
+	constructor(config: StorageConfig) {
+		this.bucketName = config.bucketName;
+		this.region = config.region;
+		this.expiresIn = config.expiresIn || EXPIRE_TIME;
+		this.endpoint = config.endpoint;
+
+		graphweaverMetadata.addMutation({
+			name: 'getUploadUrl',
+			getType: () => GraphQLJSON,
+			resolver: this.getUploadUrl.bind(this),
+			args: {
+				key: String,
+			},
+		});
+
+		graphweaverMetadata.addMutation({
+			name: 'getDeleteUrl',
+			getType: () => String,
+			resolver: this.getDeleteUrl.bind(this),
+			args: {
+				key: String,
+			},
+		});
+
+		graphweaverMetadata.addQuery({
+			name: 'getDownloadUrl',
+			getType: () => String,
+			resolver: this.getDownloadUrl.bind(this),
+			args: {
+				key: String,
+			},
+		});
+	}
+
+	async getUploadUrl(
+		_source: Source,
+		{ key }: { key: string },
+		_ctx: BaseContext,
+		_info: GraphQLResolveInfo
+	): Promise<{ url: string; filename: string; type: MediaType }> {
+		if (!this.bucketName) throw new Error('Missing required env AWS_S3_BUCKET');
```

Comment:
There are no hook calls or ACL checks on these mutations and query.

Should we add them as part of this work or create a new task to track it?

Reply:
Let's go ahead and finish this properly as part of this work.

Reply:
Captured as part of https://exogee.atlassian.net/browse/EXOGW-355